File: golang-backend/internal/handlers/projecthandler/dto.go
package projecthandler

import (
	"github.com/aportela/doneo/internal/handlers"
	"github.com/aportela/doneo/internal/handlers/projectpriorityhandler"
	"github.com/aportela/doneo/internal/handlers/projectstatushandler"
	"github.com/aportela/doneo/internal/handlers/projecttypehandler"
	"github.com/aportela/doneo/internal/handlers/userhandler"
)

type projectType struct {
	ID string `json:"id"`
}

type projectPriority struct {
	ID string `json:"id"`
}

type projectStatus struct {
	ID string `json:"id"`
}

type addRequest struct {
	ID          string          `json:"id"`
	Slug        string          `json:"slug"`
	Summary     string          `json:"summary"`
	Description *string         `json:"description"`
	Type        projectType     `json:"type"`
	Priority    projectPriority `json:"priority"`
	Status      projectStatus   `json:"status"`
}

type updateRequest struct {
	ID          string          `json:"id"`
	Slug        string          `json:"slug"`
	Summary     string          `json:"summary"`
	Description *string         `json:"description"`
	Type        projectType     `json:"type"`
	Priority    projectPriority `json:"priority"`
	Status      projectStatus   `json:"status"`
	StartedAt   *int64          `json:"startedAt"`
	FinishedAt  *int64          `json:"finishedAt"`
	DueAt       *int64          `json:"dueAt"`
}

type filterRequest struct {
	Slug            *string                   `json:"slug"`
	Summary         *string                   `json:"summary"`
	TypeID          *string                   `json:"typeId"`
	PriorityID      *string                   `json:"priorityId"`
	StatusID        *string                   `json:"statusId"`
	CreatedAt       *handlers.TimestampFilter `json:"createdAt"`
	CreatedByUserID *string                   `json:"createdByUserId"`
}

type searchRequest struct {
	Pager  handlers.PagerRequest `json:"pager"`
	Order  handlers.OrderRequest `json:"order"`
	Filter *filterRequest        `json:"filter"`
}

type projectResponse struct {
	ID                     string                                         `json:"id"`
	SLug                   string                                         `json:"slug"`
	Summary                string                                         `json:"summary"`
	Description            *string                                        `json:"description"`
	CreatedBy              userhandler.UserBaseResponse                   `json:"createdBy"`
	CreatedAt              int64                                          `json:"createdAt"`
	UpdatedAt              *int64                                         `json:"updatedAt"`
	DeletedAt              *int64                                         `json:"deletedAt"`
	StartedAt              *int64                                         `json:"startedAt"`
	FinishedAt             *int64                                         `json:"finishedAt"`
	DueAt                  *int64                                         `json:"dueAt"`
	Type                   projecttypehandler.ProjectTypeResponse         `json:"type"`
	Priority               projectpriorityhandler.ProjectPriorityResponse `json:"priority"`
	Status                 projectstatushandler.ProjectStatusResponse     `json:"status"`
	TasksCount             uint16                                         `json:"tasksCount"`
	PermissionsCount       uint16                                         `json:"permissionsCount"`
	AttachmentsCount       uint16                                         `json:"attachmentsCount"`
	NotesCount             uint16                                         `json:"notesCount"`
	HistoryOperationsCount uint16                                         `json:"historyOperationsCount"`
}

type searchProjectsResponse struct {
	Projects []projectResponse      `json:"projects"`
	Pager    handlers.PagerResponse `json:"pager"`
}

File: vite-vue-ts-frontend/src/modules/projects/types/dto.ts
import type { PagerQuery, PagerResult } from "../../../shared/types/pager";
import type { Order } from "../../../shared/types/order";

import type { ProjectTypeResponse } from "../../project-types/types/dto";
import type { ProjectPriorityResponse } from "../../project-priorities/types/dto";
import type { ProjectStatusResponse } from "../../project-statuses/types/dto";
import type { UserBaseResponse } from "../../users/types/dto";
import type { TimestampRange } from "../../../shared/composables/timestamps";
import type { AllowedProjectOperations } from "../../../shared/types/dto/allowed-project-operations";

export type AddRequest = {
  slug: string;
  summary: string;
  description: string | null;
  type: {
    id: string;
  };
  priority: {
    id: string;
  };
  status: {
    id: string;
  };
};

export type UpdateRequest = {
  id: string;
  slug: string;
  summary: string;
  description: string | null;
  type: {
    id: string;
  };
  priority: {
    id: string;
  };
  status: {
    id: string;
  };
  startedAt: number | null;
  finishedAt: number | null;
  dueAt: number | null;
};

export type PatchRequest = {
  id: string;
  status: {
    id: string;
  };
};

type SearchRequestFilter = {
  slug?: string;
  summary?: string;
  typeId?: string;
  priorityId?: string;
  statusId?: string;
  createdByUserId?: string;
  createdAt?: TimestampRange;
  updatedAt?: TimestampRange;
  finishedAt?: TimestampRange;
  dueAt?: TimestampRange;
  archivedAt?: TimestampRange;
};

export type SearchRequest = {
  pager: PagerQuery;
  order: Order;
  filter?: SearchRequestFilter;
};

export type ProjectResponse = {
  id: string;
  slug: string;
  summary: string;
  description: string;
  type: ProjectTypeResponse;
  priority: ProjectPriorityResponse;
  status: ProjectStatusResponse;
  createdAt: number;
  createdBy: UserBaseResponse;
  updatedAt?: number;
  startedAt?: number;
  finishedAt?: number;
  dueAt?: number;
  archivedAt?: number;
  tasksCount: number;
  permissionsCount: number;
  attachmentsCount: number;
  notesCount: number;
  pagesCount: number;
  historyOperationsCount: number;
  allowedOperations: AllowedProjectOperations;
};

export type SearchResponse = {
  projects: ProjectResponse[];
  pager: PagerResult;
};
